File: STRING_PY/BMH_search.py
"""Алгоритм поиска подстроки - Бойера-Мура-Хорспула"""
# Этап 1: формирование таблицы смещений
def table_shift(pattern):
    unique_symbol = set()           # уникальные символы в образе
    count_symbol = len(pattern)     # число символов в образе
    dict_shift = dict()             # словарь смещений

    for i in range(count_symbol-2, -1, -1):         # итерации с предпоследнего символа
        if pattern[i] not in unique_symbol:         # если символ еще не добавлен в таблицу
            dict_shift[pattern[i]] = count_symbol - i - 1
            unique_symbol.add(pattern[i])
    # отдельно формируем последний символ
    if pattern[count_symbol - 1] not in unique_symbol:
        dict_shift[pattern[count_symbol - 1]] = count_symbol
    # смещения для прочих символов
    dict_shift['*'] = count_symbol
    return dict_shift

# Этап 2: поиск образа в строке
def bmh_search(string, pattern):
    size_txt = len(string)
    size_pat = len(pattern)
    shift_tb = table_shift(pattern)     # {'e': 1, 's': 2, '*': 2} - таблица сдвигов

    if size_txt >= size_pat:
        i = size_pat-1                  # счетчик проверяемого символа в строке
        while(i < size_txt):
            k = 0
            j = 0
            fl_break = False
            for j in range(size_pat-1, -1, -1):
                if string[i - k] != pattern[j]:
                    print(f"i == {i}")
                    print(f"j == {j}")
                    print(f"k == {k}")

                    if j == size_pat-1:
                        #print(string[i - k])
                        # смещение, если не равен последний символ образа
                        off = shift_tb[string[i]] if shift_tb.get(string[i], False) else shift_tb['*']
                        print(f"shift_tb.get(string[i], False) == {shift_tb.get(string[i], False)}")
                    else:
                        off = shift_tb[pattern[j]]          # смещение, если не равен не последний символ образа

                    i += off                                # смещение счетчика строки
                    fl_break = True                         # если несовпадение символа, то fl_break = True
                    break
                k += 1                                      # смещение для сравниваемого символа в строке
            if not fl_break:                                # если дошли до начала образа, значит, все его символы совпали
                return i - k + 1
    return -1

def main():
    #string_1  = "The Boyer–Moore–Horspool is an algorithm for finding substrings in strings"
    string_1  = "strings"
    string_2  = "Алгоритм Бойера–Мура–Хорспула - это алгоритм для поиска подстрок в строках"
    pattern_1 = "in"
    pattern_2 = "поиска"
    print(bmh_search(string_1, pattern_1))
    #print(bmh_search(string_2, pattern_2))

if __name__ == "__main__":
    main()

File: STRING_CPP/BMH_search.cpp
#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
using namespace std;

/*****Алгоритм поиска подстроки - Бойера-Мура-Хорспула*****/
// Этап 1: формирование таблицы смещений
map<char, int64_t> table_shift(string pattern) {
    int64_t count_symbol = pattern.size();                      // Количество символов в искомой строке             
    set<char> unique_symbol;                                    // Уникальные символы в образе
    map<char, int64_t> dict_shift;                              // Словарь смещений вида {'a': 1, 'b': 2, 'c': 3, '$':3}

    for (int64_t i = count_symbol - 2; i > -1; --i) {           // Начиная с предпоследнего элемента в искомой строке
        if (unique_symbol.count(pattern[i]) != 1) {             // Если символ ещё не записан в словарь смещений
            dict_shift[pattern[i]] = count_symbol - i - 1;      // Добавить этот символ со значением индекса считая с конца
            unique_symbol.insert(pattern[i]);                   // Добавить уникальный символ в множество (в набор)
        }
    }
    if (!unique_symbol.count(pattern[count_symbol - 1])) {      // Проверить последний символ, если его нет в множестве
        dict_shift[pattern[count_symbol - 1]] = count_symbol;   // Тогда добавить символ в словарь зо значением последнего индекса + 1
    }
    dict_shift['%'] = count_symbol;                             // Добавить в словарь ограничительный символ
    return dict_shift;                                          // Вернуть получившийся словарь "смещений"
}

/**************Сам алгоритм поиска - bmh_search***************/
// Этап 2: поиск образа в строке
int64_t bmh_search(string txt, string pat) {               
    int64_t size_txt = static_cast<int64_t>(txt.size());        // Размер исходной строки
    int64_t size_pat = static_cast<int64_t>(pat.size());        // Размер искомой строки
    int64_t i = 0, k = 0, j = 0, off = 0;                       // Индексы для манипуляций в цикле
    map<char, int64_t> shift_tb = table_shift(pat);             // Словарь смещений сформированный по искомой строке

    if (size_txt >= size_pat) {                                 // Если исходная меньше искомой - вернуть -1
        i = size_pat - 1;                                       // Потск начинается с этого места
        while (i < size_txt) {                                  // Пока индекс меньше длины шаблона итерироваться по строке
            k = j = 0;                                          // Опять копипаста (можно опустить)
            bool fl_break = false;                              // Флаг для прерыванияцикла, если найдется подстрока или превысит размер
            for (j = size_pat - 1; j >= 0; --j) {               // Начиная с индекса == длине шаблона и в обратную сторону
                if (txt[i - k] != pat[j]) {                     // Если указатели не вышли за предел строки исходной, то
                    if (j == size_pat - 1) {                    // Проверить равен индекс длине шаблона
                        // Если равен тогда проверить посимвольно
                        off = shift_tb.count(txt[i]) != 1 ? shift_tb['%'] : shift_tb[txt[i]];
                    }
                    else {
                        off = shift_tb[pat[j]];
                    }
                    i = i + off;
                    fl_break = true;
                    break;
                }
                k++;
            }
            if (!fl_break) {
                return i - k + 1;
            }
        }
    }
    return -1;
}


int main() {
    setlocale(0, "Ru");
    string en_str = "The Boyer–Moore–Horspool is an algorithm for finding substrings in strings";
    string ru_str = "Алгоритм Бойера–Мура–Хорспула - это алгоритм для поиска подстрок в строках";
    string en_pat = "rithm";
    string ru_pat = "поиск";
    cout << "En pattern index == " << bmh_search(en_str, en_pat) << "\n";
    cout << "========================================================\n";
    cout << "Ru pattern index == " << bmh_search(ru_str, ru_pat) << "\n";
    return 0;
}